src/Microsoft.VisualStudio.LanguageServices.Razor/Editor/RazorDirectiveCompletionProvider.cs

```diff
@@ -0,0 +1,185 @@
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+
+            if (!AtDirectiveCompletionPoint(syntaxTree, context))
+            {
+                // Can't have a valid directive at the current location.
+                return Task.FromResult(0);
+            }
+
+            var directives = syntaxTree.Options.Directives.Concat(DefaultDirectives);
+            foreach (var directive in directives)
+            {
+                var propertyDictionary = new Dictionary<string, string>(StringComparer.Ordinal);
+
+                if (!string.IsNullOrEmpty(directive.Description))
+                {
+                    propertyDictionary[DescriptionKey] = directive.Description;
+                }
+
+                var completionItem = CompletionItem.Create(
+                    directive.Directive,
+                    // This groups all Razor directives together
+                    sortText: "_RazorDirective_",
+                    rules: CompletionItemRules.Create(formatOnCommit: false),
+                    tags: ImmutableArray.Create(CompletionTags.Intrinsic),
+                    properties: propertyDictionary.ToImmutableDictionary());
+                context.AddItem(completionItem);
+            }
+
+            return Task.FromResult(0);
+        }
+
+        private bool AtDirectiveCompletionPoint(RazorSyntaxTree syntaxTree, CompletionContext context)
+        {
+            if (TryGetRazorSnapshotPoint(context, out var razorSnapshotPoint))
+            {
+                var change = new SourceChange(razorSnapshotPoint.Position, 0, string.Empty);
+                var owner = syntaxTree.Root.LocateOwner(change);
+                if (owner.ChunkGenerator is ExpressionChunkGenerator &&
+                    owner.Symbols.All(IsDirectiveCompletableSymbol) &&
+                    // Do not provide IntelliSense for explicit expressions. Explicit expressions will usually look like:
+                    // [@] [(] [DateTime.Now] [)]
+                    owner.Parent?.Children.Count > 1 &&
+                    owner.Parent.Children[1] == owner)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        protected virtual bool TryGetRazorSnapshotPoint(CompletionContext context, out SnapshotPoint snapshotPoint)
+        {
+            snapshotPoint = default(SnapshotPoint);
+
+            if (context.Document.TryGetText(out var sourceText))
+            {
+                var textSnapshot = sourceText.FindCorrespondingEditorTextSnapshot();
+                var projectionSnapshot = textSnapshot as IProjectionSnapshot;
+
+                if (projectionSnapshot == null)
+                {
+                    return false;
+                }
+
+                var mappedPoints = projectionSnapshot.MapToSourceSnapshots(context.CompletionListSpan.Start);
+                var htmlSnapshotPoint = mappedPoints
+                    .Cast<SnapshotPoint?>()
```

Comment:
do we need a null check since this is nullable?

Reply:
Going to change this logic slightly.